duouniversal/client.go

```diff
@@ -218,28 +218,60 @@ func newStrictTLSTransport() *http.Transport {
 	}
 }
 
-func NewClient(clientId, clientSecret, apiHost, redirectUri string) (*Client, error) {
-	return NewClientDuoCodeAttribute(clientId, clientSecret, apiHost, redirectUri, true)
+func NewClient(clientId, clientSecret, apiHost, redirectUri string, opts ...clientOptions) (*Client, error) {
+	opts = append(opts, UseDuoCodeAttribute())
+	return newClient(clientId, clientSecret, apiHost, redirectUri, opts...)
 }
 
 // Creates a new Client with the ability to turn off use_duo_code_attribute
 func NewClientDuoCodeAttribute(clientId, clientSecret, apiHost, redirectUri string, useDuoCodeAttribute bool) (*Client, error) {
-	if len(clientId) != clientIdLength {
-		return nil, fmt.Errorf(clientIdError)
-	} else if len(clientSecret) != clientSecretLength {
-		return nil, fmt.Errorf(clientSecretError)
-	}
-
-	return &Client{
-		clientId:            clientId,
-		clientSecret:        clientSecret,
-		apiHost:             apiHost,
-		redirectUri:         redirectUri,
-		useDuoCodeAttribute: useDuoCodeAttribute,
-		duoHttpClient: &http.Client{
-			Transport: newStrictTLSTransport(),
-		},
-	}, nil
+    if useDuoCodeAttribute {
+        return newClient(clientId, clientSecret, apiHost, redirectUri, UseDuoCodeAttribute())
+    }
+    return newClient(clientId, clientSecret, apiHost, redirectUri)
+}
+
+// Creates a new Client with the ability to turn off use_duo_code_attribute
+func newClient(clientId, clientSecret, apiHost, redirectUri string, opts ...clientOptions) (*Client, error) {
+    if len(clientId) != clientIdLength {
+        return nil, fmt.Errorf(clientIdError)
+    }
+    if len(clientSecret) != clientSecretLength {
+        return nil, fmt.Errorf(clientSecretError)
+    }
+
+    c := &Client{
+        clientId:            clientId,
+        clientSecret:        clientSecret,
+        apiHost:             apiHost,
+        redirectUri:         redirectUri,
+        useDuoCodeAttribute: false,
+        duoHttpClient: &http.Client{
+            Transport: newStrictTLSTransport(),
+        },
+    }
+
+    for _, option := range opts {
```

Comment:
I believe there's no mechanism for limiting the possible options to only the ones created below?  i.e. anyone could create a method that has the same signature and pass it in, and it will get executed?  If that's the case, is it worth formalizing an interface (if Go has interfaces?)

Reply:
Thanks, it's not a showstopper; this is just another place I want to understand the ramifications better.